lib/spack/spack/paths.py

```diff
@@ -50,7 +50,12 @@
 mock_packages_path = os.path.join(repos_path, "builtin.mock")
 
 #: User configuration location
-user_config_path = os.path.expanduser('~/.spack')
+__SPACK_USER_CONFIG = os.environ.get('SPACK_USER_CONFIG')
```

Comment:
I'm not sure that this should be environment dependent, why is it insufficient to set it via `etc/config.yaml`?

Reply:
Is it currently possible to set it in etc/config.yaml?  My understanding was that `~/.spack` is used unconditionally at present.  Having it be a setting you can get from either etc, or from an environment, would be great and I wouldn't mind making that change, but wanted to get this up here to spark this kind of discussion.  Thoughts @alalazo, @tgamblin?

Reply:
Yes you are correct `~/.spack` is currently unconditionally used. This bit us multiple times (... we (mostly) amputated the user config scope) and I love to have it configurable.

But I'd prefer to have it settable via the config file rather than the environment (I have to adapt the config anyways and the environment is uncontrolled, from the perspective of the spack repo).

Reply:
I largely agree, but we have a control inversion, or at least ordering, problem. In principle the user config overrides the environment, repo and system configs.  I'm thinking we have a few options:

* add a way  to configure how spack should order configs, something like XDG_CONFIG_DIRS, so you can order how all configurations should be applied, example: `$env:$repo:$user:/some/local/config/path:$system` to give env highest priority over repo and user
  * Apply this with an env var
  * Apply this with a new file in home or XDG_CONFIG_HOME or similar, that can be used to set up environment before any normal config is used, think .zshrc loading before the zsh config path is searched for .zshenv so .zshenv can live anywhere
* make this setting, or a file containing it, "special" so that one can make a truly local repo without need for environment or user config to say so

Reply:
Having the value of the user config path settable in `etc/config.yaml` will likely not work with a shared instance of Spack since not all users will have write privileges.  The environment variable seems the easiest way to implement this behavior.

Reply:
These lines could be simplified as:

```python
user_config_path = os.getenv("SPACK_USER_CONFI_PATH", os.path.expanduser("~/.spack"))
```

Reply:
Good point, that's actually not quite right but moving the expanduser outside the getenv does what I want, thanks @tjfulle!

As to `etc/config.yaml`, I agree the env var is most flexible for a user of a shared spack, but it means that the owner of that specific installation of spack can't set a default for the usage of it.  I can see uses either direction, but the env var is simpler so going for that first makes a lot of sense to me.